### PR TITLE
Add Rich Nav specific pipeline, temporarily add to ci for testing

### PR DESCRIPTION
**Pull request by jepetty**
I've altered two files:
1. Add a Rich Navigation specific build (which will appear as a new build pipeline), indexing on every push to master. This will allow you to use Rich Navigation in Visual Studio to enable cross-repository searches, serverless navigation at https://online.visualstudio.com/github/dotnet/msbuild, and integrated GitHub C# navigation once that is released (or, comment below if you want to be added to the preview now 😊).
1. Added the Rich Navigation task to your normal CI. This has all the same benefits of ^, but also indexes your pull requests for a richer code review experience. This does come at the cost of increasing the time of your CI.

Based on the size of the repository and frequency of changes being pushed, I highly recommend we move forward with Option 1 and create a new build specifically for Rich Navigation (this is what we did in dotnet/roslyn as well)

## Rich Code Navigation

[Docs Link](https://devdiv.visualstudio.com/DevDiv/_wiki/wikis/DevDiv.wiki/1047/Rich-Code-Navigation)
With Rich Code Navigation, you can use editor-level navigation capabilities (like peek definition, find all references, and even diagnostics) on a pull request, without requiring a local checkout. This is available across all files and dependencies of your repository.

Rich Code Navigation is available for Visual Studio (for repos hosted on GitHub or Azure Repos) and Visual Studio Code (for repos hosted on GitHub).

Rich Code Navigation supports these capabilities:

- Hover
- Peek definition and Go to definition
- Peek references and Find all references
- Go to type definition
- Go to implementation
- Diagnostics

### CODE REVIEW

File: .azure/pipelines/ci.yml
Comment (dougbu) [inline .azure/pipelines/ci.yml:85]: Please revert changes to our regular CI. There's an error in the file now in any case.

File: .azure/pipelines/richnav.yml
Comment (dougbu) [inline .azure/pipelines/richnav.yml:41]: Remove all unused variables. Suspect only `$(Windows64LogArgs)` `$(Windows86LogArgs)` and `$(WindowsInstallersLogArgs)` are needed in the list after `$(_BuildArgs)`

File: .azure/pipelines/richnav.yml
Comment (dougbu) [inline .azure/pipelines/richnav.yml:59]: This comment is not applicable for this pipeline

File: .azure/pipelines/richnav.yml
Comment (dougbu) [inline .azure/pipelines/richnav.yml:66]: Remove references to the empty `$(_InternalRuntimeDownloadArgs)` variable

File: .azure/pipelines/ci.yml
Comment (jepetty) [reply .azure/pipelines/ci.yml:85]: Will do - I never intended to merge these changes, just made these changes to get a glimpse of what the time to run would be and to verify the rich nav pipeline would run (I can't push my branch to the upstream source so I can't make a new pipeline off my branch until after it's merged... it makes me have to do some juggling to validate confidently that things will work as expected 😃 )

**Comment by jepetty** [comment]
This merge should also not complete until https://github.com/dotnet/arcade/pull/6345 completes with fixes to the CI templates

File: .azure/pipelines/richnav.yml
Comment (dougbu) [inline .azure/pipelines/richnav.yml:34]: Don't we need the following here too❔
``` text
richCodeNavigationLanguage: 'csharp,typescript,java'
```

Comment (dougbu) [reply .azure/pipelines/richnav.yml:34]: Oh, never mind.

**Comment by dougbu** [comment]
> This merge should also not complete until dotnet/arcade#6345 completes with fixes to the CI templates

We keep an eye on arcade promotions and can trigger the right subscriptions (for this repo, aspnetcore, and aspnetcore-tooling) once your change is in.

Do you also need to test w/ the newly-added pipeline before we merge this❔

**Comment by jepetty** [comment]
Definitely, Doug! I'll link to a successful build before I merge. Thanks for the review and help!

**Comment by dougbu** [comment]
/fyi @jepetty #27034 will bring your Arcade changes into this repo. Any harm in merging this before that finishes up❔

**Comment by jepetty** [comment]
Please ignore my last hack-tacular commit to replace the regular CI with the Rich Nav pipeline. It's only temporary while I confirm that the build succeeds, then I'll revert the change

**Comment by dougbu** [comment]
@jepetty we absolutely do _not_ want to take a change that turns on `$(DisablePackageReferenceRestrictions)` globally. What makes you think this is needed for the rich nav experience❔

**Comment by jepetty** [comment]
@dougbu we are currently adding the RichCodeNav.EnvVarDump package as a PackageReference in dotnet/arcade, so without `$(DisablePackageReferenceRestrictions)` the build fails (like [here](https://dnceng.visualstudio.com/public/_build/results?buildId=870514&view=logs&j=1b89928a-2219-5ef9-602f-f95beb3da4dc&t=2f754404-b52c-5665-45a1-cdacbbf9ac37)). I had (possibly wrongly :D) assumed that it would be ok in just the Rich Nav pipeline, but I obviously don't know the intricacies of this build system.

I can test and update the reference addition in the arcade repo, it'll just take some time to propagate. Do you have any other suggestions though?

**Comment by dougbu** [comment]
@jepetty update the `@(PackageReference)` item in Arcade to include `IsImplicitlyDefined="true"` i.e.
``` diff
- <PackageReference Include="RichCodeNav.EnvVarDump" Version="$(RichCodeNavPackageVersion)" PrivateAssets="all" />
+ <PackageReference Include="RichCodeNav.EnvVarDump" Version="$(RichCodeNavPackageVersion)" IsImplicitlyDefined="true" PrivateAssets="all" />
```

Arcade already does this a lot of other places.

**Comment by jepetty** [comment]
Ok fantastic, thanks for the insight! I've removed the portion setting `$(DisablePackageReferenceRestrictions)` to true, and started a [pull request](https://github.com/dotnet/arcade/pull/6513) to update the arcade repo with your recommendation

File: NuGet.config
Comment (dougbu) [inline NuGet.config:16]: @jepetty @mmitche could this package be included in dotnet-eng instead❔ Seems odd for us to need a new feed for something referenced from the Arcade SDK.

File: .azure/pipelines/richnav.yml
Comment (dougbu) [inline .azure/pipelines/richnav.yml:63]: This step may not be worthwhile unless we can include C++ in the rich nav languages. The Windows installers build only includes 2 C# projects.

File: .azure/pipelines/richnav.yml
Comment (dougbu) [inline .azure/pipelines/richnav.yml:16]: If this were left out, would the test code get rich nav too❔ Might be worthwhile.

File: .azure/pipelines/jobs/default-build.yml
Comment (dougbu) [inline .azure/pipelines/jobs/default-build.yml:82]: Curious: Any other supported languages e.g. this repo contains JavaScript, C++, and a bit of F#❔

**Review by dougbu — APPROVED** [review]
Looks good. Can make further improvements separately&hellip;

**Comment by Pilchie** [comment]
@dougbu - should we go ahead and merge this?

**Comment by dougbu** [comment]
> should we go ahead and merge this?

I think so but wanted @jepetty to confirm she also thought this was ready. @jepetty❔ 

(Will merge over the weekend if we don't hear back.)

**Pull request merged**
